lib/collection/src/collection_manager/optimizers/vacuum_optimizer.rs

```diff
@@ -64,26 +69,107 @@ impl VacuumOptimizer {
+                let littered_ratio_segment = self.littered_ratio_segment(segment);
+                let littered_ratio_vectors = self.littered_ratio_vectors(segment);
+                [littered_ratio_segment, littered_ratio_vectors]
+                    .into_iter()
+                    .flatten()
+                    .map(|ratio| (*idx, ratio))
+            })
+            .max_by_key(|(_, ratio)| OrderedFloat(*ratio))
+            .map(|(idx, _)| (idx, segments_read_guard.get(idx).unwrap().clone()))
+    }
 
-                let segment_entry = segment.get();
-                let read_segment = segment_entry.read();
-                let littered_ratio =
-                    read_segment.deleted_point_count() as f64 / read_segment.points_count() as f64;
+    /// Calculate littered ratio for segment on point level
+    ///
+    /// Returns `None` if littered ratio did not reach vacuum thresholds.
+    fn littered_ratio_segment(&self, segment: &LockedSegment) -> Option<f64> {
+        let segment_entry = segment.get();
+        let read_segment = segment_entry.read();
 
-                let is_big = read_segment.points_count() >= self.min_vectors_number;
-                let is_not_special = read_segment.segment_type() != SegmentType::Special;
-                let is_littered = littered_ratio > self.deleted_threshold;
+        // Never optimize special segments
+        if read_segment.segment_type() == SegmentType::Special {
+            return None;
+        }
+
+        let littered_ratio =
+            read_segment.deleted_point_count() as f64 / read_segment.total_point_count() as f64;
+        let is_big = read_segment.total_point_count() >= self.min_vectors_number;
+        let is_littered = littered_ratio > self.deleted_threshold;
 
-                (is_big && is_not_special && is_littered).then_some((*idx, littered_ratio))
+        (is_big && is_littered).then_some(littered_ratio)
+    }
+
+    /// Calculate littered ratio for segment on vector level
+    ///
+    /// If a segment has multiple named vectors, it checks each one.
+    ///
+    /// The ratio is based on the number of deleted vectors in each named vector in which the
+    /// vacuum threshold is reached.
+    ///
+    /// Returns `None` if littered ratio did not reach vacuum thresholds for no named vectors.
+    fn littered_ratio_vectors(&self, segment: &LockedSegment) -> Option<f64> {
+        let segment_entry = segment.get();
+        let read_segment = segment_entry.read();
+
+        // Never optimize special segments
+        if read_segment.segment_type() == SegmentType::Special {
+            return None;
+        }
+
+        // Segment must have an index
+        let segment_config = read_segment.config();
+        if !segment_config.is_vector_indexed() {
+            return None;
+        }
+
+        // We can only work with original segments
+        let real_segment = match segment {
+            LockedSegment::Original(segment) => segment.read(),
```

Comment:
Double read: `segment_entry.read();`  and `segment.read()` is the same mutex. Might result in a deadlock here